pkg/mounter/safe_mounter_windows.go

```diff
@@ -113,13 +124,54 @@ func (mounter *csiProxyMounter) SMBMount(source, target, fsType string, mountOpt
 		return fmt.Errorf("smb mapping failed with error: %v", err)
 	}
 	klog.V(2).Infof("mount %s on %s successfully", source, normalizedTarget)
+
+	if mappingPath != "" {
+		// When mount successful tries to increment mapping reference count.
+		err = incementMappingPathCount(mappingPath, source)
+		if err != nil {
+			klog.Warningf("incementMappingPathCount(%s, %s) failed with %v", mappingPath, source, err)
+		}
+	}
+
 	return nil
 }
 
 func (mounter *csiProxyMounter) SMBUnmount(target string) error {
 	klog.V(4).Infof("SMBUnmount: local path: %s", target)
-	// TODO: We need to remove the SMB mapping. The change to remove the
-	// directory brings the CSI code in parity with the in-tree.
+
+	if remotePath, err := os.Readlink(target); err != nil {
+		klog.Warningf("SMBUnmount: can't get remote path: %v", err)
+	} else {
+		mappingPath, err := getRootMappingPath(remotePath)
+		if err != nil {
+			klog.Warningf("getRootMappingPath(%s) failed with %v", remotePath, err)
+		} else {
+			klog.V(4).Infof("SMBUnmount: remote path: %s, mapping path: %s", remotePath, mappingPath)
+
+			unlock := lock(mappingPath)
+			defer unlock()
+
+			if err := decrementMappingPathCount(mappingPath, remotePath); err != nil {
+				klog.Warningf("decrementMappingPathCount(%s, %d) failed with %v", mappingPath, remotePath, err)
+			} else {
+				count := getMappingPathCount(mappingPath)
+				if count == 0 {
+					smbUnmountRequest := &smb.RemoveSmbGlobalMappingRequest{
+						RemotePath: remotePath,
+					}
+					klog.V(2).Infof("begin to unmount %s on %s", remotePath, target)
+					if _, err := mounter.SMBClient.RemoveSmbGlobalMapping(context.Background(), smbUnmountRequest); err != nil {
+						klog.Warningf("smb unmapping failed with error: %v", err)
```

Comment:
I would suggest return error here, it's now silent failure:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-csi_csi-driver-smb/505/pull-csi-driver-smb-e2e-windows-containerd/1555996346316492800/build-log.txt
```
[pod/csi-smb-node-win-dv6m9/smb] I0806 19:50:13.988175    6564 safe_mounter_windows.go:162] begin to unmount \\10.0.255.191\share\pvc-33bb1b4f-930f-484e-8733-e952be2a9da0\ on \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-33bb1b4f-930f-484e-8733-e952be2a9da0\globalmount
[pod/csi-smb-node-win-dv6m9/smb] W0806 19:50:14.831750    6564 safe_mounter_windows.go:164] smb unmapping failed with error: rpc error: code = Unknown desc = UnmountSmbShare failed. output: "Remove-SmbGlobalMapping : No MSFT_SmbGlobalMapping objects found with property 'RemotePath' equal to \r\n'\\\\10.0.255.191\\share\\pvc-33bb1b4f-930f-484e-8733-e952be2a9da0\\'.  Verify the value of the property and retry.\r\nAt line:1 char:1\r\n+ Remove-SmbGlobalMapping -RemotePath $Env:smbremotepath -Force\r\n+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : ObjectNotFound: (\\\\10.0.255.191\\...3-e952be2a9da0\\:String) [Remove-SmbGlobalMapping], Ci \r\n   mJobException\r\n    + FullyQualifiedErrorId : CmdletizationQuery_NotFound_RemotePath,Remove-SmbGlobalMapping\r\n \r\n", err: exit status 1
[pod/csi-smb-node-win-dv6m9/smb] I0806 19:50:14.831750    6564 safe_mounter_windows.go:204] Remove directory: \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-33bb1b4f-930f-484e-8733-e952be2a9da0\globalmount
```

Reply:
If we do it, then we need to make changes in other tests because now it fails when we return error. 
Is it Ok?

Reply:
I think this is a bug per below error msg:
```
smb unmapping failed with error: rpc error: code = Unknown desc = UnmountSmbShare failed. output: "Remove-SmbGlobalMapping : No MSFT_SmbGlobalMapping objects found with property 'RemotePath' equal to \r\n'\\\\10.0.255.191\\share\\pvc-33bb1b4f-930f-484e-8733-e952be2a9da0\\'
```

`RemoveSmbGlobalMapping` should work in this PR, right?

Reply:
I think the problem in `\` at end of smb share. We need to insure that remotePath has the same value when we mount/unmount. And `\` matters in this case.

Reply:
Last changes fix this issue:
```
[pod/csi-smb-node-win-8tdjm/smb] I0808 08:10:48.205840    6608 safe_mounter_windows.go:143] SMBUnmount: local path: \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d\globalmount
[pod/csi-smb-node-win-8tdjm/smb] I0808 08:10:48.205895    6608 safe_mounter_windows.go:155] SMBUnmount: remote path: \\10.0.21.74\share\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d, mapping path: \\10.0.21.74\share
[pod/csi-smb-node-win-8tdjm/smb] I0808 08:10:48.206477    6608 safe_mounter_windows.go:168] begin to unmount \\10.0.21.74\share\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d on \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d\globalmount
[pod/csi-smb-node-win-8tdjm/smb] I0808 08:10:48.948243    6608 safe_mounter_windows.go:172] unmount \\10.0.21.74\share\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d on \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d\globalmount successfully
[pod/csi-smb-node-win-8tdjm/smb] I0808 08:10:48.948321    6608 safe_mounter_windows.go:210] Remove directory: \var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-e75bec88-9d12-440c-baa6-3f46cf76d32d\globalmount
```

Reply:
thanks, I think we should return error in L163, and then squash all commits?

Reply:
done